File: Graph/743. network-delay-time.cpp
/*
https://leetcode.com/problems/network-delay-time/

There are N network nodes, labelled 1 to N.
Given times, a list of travel times as directed edges times[i] = (u, v, w), where u is the source node,
v is the target node, and w is the time it takes for a signal to travel from source to target.
Now, we send a signal from a certain node K. How long will it take for all nodes to receive the signal?
If it is impossible, return -1.

Example 1:
Input: times = [[2,1,1],[2,3,1],[3,4,1]], N = 4, K = 2
Output: 2
 
Note:
N will be in the range [1, 100].
K will be in the range [1, N].
The length of times will be in the range [1, 6000].
All edges times[i] = (u, v, w) will have 1 <= u, v <= N and 0 <= w <= 100.
*/

class Solution {
public:

	vector<pair<int,int>> *graph;

	int networkDelayTime(vector<vector<int>>& times, int N, int source) {
		// build graph
		// since given graph is 1 index based, so convert it to
		// zero index based (reason for source-1)
		graph = new vector<pair<int,int>>[N];
		for (auto i: times) {
			int u = i[0]-1, v = i[1]-1, weight = i[2];
			graph[u].push_back({v, weight});
		}
		// return bellmanFord(N, source-1);
		return dijkstra(N, source-1);
	}
	
	// ********** SLOWER ***********
	int bellmanFord(int N, int source) {
		vector<int> dist(N, INT_MAX); // min dist array
		dist[source]=0;
		// to update min distance {a path can have at most N-1 edges}
		for (int pathLenght=1; pathLenght<=N-1; pathLenght++) {
			for (int i=0; i<N; i++) { // explore all edges
				vector<pair<int,int>>& curr = graph[i];
				for (auto j:curr) {
					int u=i, v=j.first, weight=j.second;
					if (dist[u]!=INT_MAX and dist[u]+weight < dist[v])
						dist[v] = dist[u]+weight;
				}
			}
		}
		int ans = -1;
		for (auto i:dist) ans = max(ans, i);
		return (ans==INT_MAX) ? -1 : ans;
	}

	// ********** FASTER ***********
	int dijkstra(int N, int source) {
		vector<bool> finalized(N, false); // whether min dist is finalized for ith node
		vector<int> minDistArray(N, INT_MAX); // min dist array
		minDistArray[source] = 0; // dist to source is 0
		for (int count=0; count<N-1; count++) {
			// pick min dist vertex that isn't yet finalized
			int u = minDistance(finalized, minDistArray);
			finalized[u] = true; // mark finalized
			for (auto i:graph[u]) {
				int v = i.first, weight = i.second;
				if (!finalized[v]) // if not finalized, update min distance
					minDistArray[v] = min(minDistArray[v], minDistArray[u]+weight);
			}
		}
		int ans = -1;
		for (auto i:minDistArray) ans = max(ans, i);		
		return (ans==INT_MAX) ? -1 : ans;
	}

	int minDistance(vector<bool>& finalized, vector<int>& minDistArray) {
		int index, minDist = INT_MAX;
		for (int i=0; i<finalized.size(); i++) {
			if (!(finalized[i]) and minDistArray[i]<minDist)
				minDist=minDistArray[i], index=i;
		}
		return index;
	}
};

File: Graph/743. network-delay-time.py
# https://leetcode.com/problems/network-delay-time/


class Solution:
	def networkDelayTime(self, times: List[List[int]], n: int, source: int) -> int:
		# build graph
		graph = [[] for _ in range(n)]
		for u, v, w in times:
			graph[u-1].append((v-1, w))

		# return self.bellman_ford(graph, n, source-1)
		# return self.dijkstra_ALgo(graph, n, source-1)
		return self.dijkstra_with_heap(graph, n, source-1)



	# -------------- BELLMAN FORD ----------------
	def bellman_ford(self, graph, n, source) -> int:
		dist = [float("inf")] * n
		dist[source] = 0
		# to update min distance {a path can have at most N-1 edges}
		for _ in range(n-1):
			# explore all edges
			for u in range(n):
				for v, w in graph[u]:
					if dist[u]!=float("inf") and dist[u] + w < dist[v]:
						dist[v] = dist[u] + w

		ans = max(dist)
		return ans if ans < float("inf") else -1



	# -------------- DIJKSTRA ----------------
	def dijkstra_ALgo(self, graph, n, source) -> int:
		finalized = [False] * n
		dist = [float("inf")] * n
		dist[source] = 0

		# returns min dist vertex (from source) that isn't yet finalized
		def getMinDistVertex():
			index, minDist = -1, float("inf");
			for i, d in enumerate(dist):
				if not finalized[i] and d < minDist:
					index, minDist = i, d
			return index
			# return dist.index(min([d for i, d in enumerate(dist) if not finalized[i]]))

		for _ in range(n - 1):
			# pick min dist vertex that isn't yet finalized
			u = getMinDistVertex()
			finalized[u] = True

			for v, w in graph[u]:
				if not finalized[v]:
					dist[v] = min(dist[v], dist[u] + w)

		ans = max(dist)
		return ans if ans < float("inf") else -1



	# -------------- DIJKSTRA with min Heap ----------------
	def dijkstra_with_heap(self, graph, n, source) -> int:
		dist = [float("inf")] * n
		heap = [(0, source)]

		while heap:
			d, u = heapq.heappop(heap)
			if d < dist[u]:
				dist[u] = d
				for v, w in graph[u]:
					heapq.heappush(heap, (d+w, v))

		ans = max(dist)
		return ans if ans < float("inf") else -1


















import heapq
from collections import deque
from collections import defaultdict

# """
#  Uses simple DFS - Accepted
class Solution(object):
	def networkDelayTime(self, times, N, K):
		graph = defaultdict(list)
		for u, v, w in times:
			graph[u].append((v, w))
		distance = {node: float("inf") for node in range(1, N + 1)}
		self.DFS(graph, distance, K, 0)
		totalTime = max(distance.values())
		return totalTime if totalTime < float("inf") else -1

	def DFS(self, graph, distance, node, elapsedTimeSoFar):
		if elapsedTimeSoFar >= distance[node]:   # signal aalreaady reached to this node. so no need to explore for this node
			return
		distance[node] = elapsedTimeSoFar
		for neighbour, time in sorted(graph[node]):
			self.DFS(graph, distance, neighbour, elapsedTimeSoFar + time)
# """


# """
#  Uses simple BFS - Accepted
class Solution:
	def networkDelayTime(self, times, N, K):
		elapsedTime, graph, queue = [0] + [float("inf")] * N, defaultdict(list), deque([(0, K)])
		for u, v, w in times:
			graph[u].append((v, w))
		while queue:
			time, node = queue.popleft()
			if time < elapsedTime[node]:
				elapsedTime[node] = time
				for v, w in graph[node]:
					queue.append((time + w, v))
		mx = max(elapsedTime)
		return mx if mx < float("inf") else -1
# """



# """
# Dijkstra algorithm - Accepted
class Solution:
	def networkDelayTime(self, times, N, K):
		elapsedTime, graph, heap = [0] + [float("inf")] * N, defaultdict(list), [(0, K)]
		for u, v, w in times:
			graph[u].append((v, w))
		while heap:
			time, node = heapq.heappop()
			if time < elapsedTime[node]:
				elapsedTime[node] = time
				for v, w in graph[node]:
					heapq.heappush(heap, (time + w, v))
		mx = max(elapsedTime)
		return mx if mx < float("inf") else -1
# """



# """
# Original Bellman–Ford algorithm - Accepted
class Solution:
	def networkDelayTime(self, times, N, K):
		distance = [float("inf") for _ in range(N)]
		distance[K-1] = 0
		for _ in range(N-1):
			for u, v, w in times:
				if distance[u-1] + w < distance[v-1]:
					distance[v-1] = distance[u-1] + w
		return max(distance) if max(distance) < float("inf") else -1
# """



# """
# Shortest Path Faster Algorithm (SPFA): An improvement of the Bellman–Ford algorithm - Accepted
class Solution:
	def networkDelayTime(self, times, N, K):
		elapsedTime, graph, queue = [0] + [float("inf")] * N, defaultdict(list), deque([(0, K)])
		elapsedTime[K] = 0
		for u, v, w in times:
			graph[u].append((v, w))
		while queue:
			time, node = queue.popleft()
			for neighbour in graph[node]:
				v, w = neighbour
				if time + w < elapsedTime[v]:
					elapsedTime[v] = time + w
					queue.append((time + w, v))
		mx = max(elapsedTime)
		return mx if mx < float("inf") else -1
# """





# """
#  Floyd Warshall - Accepted
class Solution:
	def networkDelayTime(self, times, N, K):
		elapsedTimeMatrix = [[float("inf") for _ in range(N)] for _ in range(N)]
		for u, v, w in times:
			elapsedTimeMatrix[u - 1][v - 1] = w
		for i in range(N):					  #   Assigning 0 to the diagonal cells
			elapsedTimeMatrix[i][i] = 0
		for k in range(N):
			for i in range(N):
				for j in range(N):
					elapsedTimeMatrix[i][j] = min(elapsedTimeMatrix[i][j], elapsedTimeMatrix[i][k] + elapsedTimeMatrix[k][j])
		mx = max(elapsedTimeMatrix[K - 1])
		return mx if mx < float("inf") else -1
# """